### PR TITLE
feat: Support --browser cli option in Launchpad

### LINKED ISSUE #18419: Add support for `cypress open --browser <browser>`
### What would you like?

```
cypress open --browser firefox
```

Launchpad opens to select testing type if not selected already and if **firefox** browser is found, preselect it.

```
cypress open --browser does-not-exist
```

Launchpad opens to select testing type. If **does-not-exist** browser is not found, error.

### Why is this needed?

_No response_

### Other

_No response_

### PR DESCRIPTION
**Pull request by chrisbreiding**
- Closes #18419 

### User facing changelog

`cypress open --browser <browser>` will automatically select the browser.

### Additional details

Displays this when the browser can't be found or isn't supported:

```
cypress open --browser does-not-exist
```

Screen Shot 2021-11-02 at 3 05 22 PM

### PR Tasks

- [ ] Have tests been added/updated?
- [ ] Has the original issue or this PR been tagged with a release in ZenHub? 
- [ ] Has a PR for user-facing changes been opened in [`cypress-documentation`](https://github.com/cypress-io/cypress-documentation)? 
- [ ] Have API changes been updated in the [`type definitions`](https://github.com/cypress-io/cypress/blob/develop/cli/types/cypress.d.ts)?
- [ ] Have new configuration options been added to the [`cypress.schema.json`](https://github.com/cypress-io/cypress/blob/develop/cli/schema/cypress.schema.json)?

### CODE REVIEW

File: packages/data-context/src/DataContext.ts
Comment (BlueWinds) [inline packages/data-context/src/DataContext.ts:71]: As we put more logic in here, I start to wonder if this function is the wrong place for some of these things.

We make a `new DataContext`, then have a bunch of promises and other functions to mutate the state - it sort of feels like this should be done beforehand, so we can initialize the DataContext in the proper state to begin with? I'm not sure exactly where it should go, but a growing list of conditionals - mostly related to CLI and environmental args - starts to feel like this is in the wrong place.

File: packages/server/lib/makeDataContext.ts
Comment (BlueWinds) [inline packages/server/lib/makeDataContext.ts:27]: Seems like rather than declaring wrappers around our imports, we could just do

```
appApi: {
  getBrowsers,
  ensureAndGetByNameOrPath,
}
```

(same elsewhere in this file)

**Review by BlueWinds — CHANGES_REQUESTED** [review]

File: packages/data-context/src/DataContext.ts
Comment (chrisbreiding) [reply packages/data-context/src/DataContext.ts:71]: Yeah, I feel the same. Thoughts, @tgriesser?

File: packages/data-context/src/actions/AppActions.ts
Comment (lmiller1990) [inline packages/data-context/src/actions/AppActions.ts:9]: I think we can just type this as `Promise<......>`. As we move forward I think we should just standardize on native promises (with which Bluebird should be compatible, anyway).

File: packages/data-context/src/actions/AppActions.ts
Comment (lmiller1990) [inline packages/data-context/src/actions/AppActions.ts:15]: Instead of making this nullable I wonder if it would be more clearer to just make the argument required and then do the check in the other function?

```ts
setActiveBrowser (browser: FoundBrowser) {
  this.ctx.coreData.wizard.chosenBrowser = browser
}

 setActiveBrowserById (id: string) {
    const browserId = this.ctx.fromId(id, 'Browser')

    // Ensure that this is a valid ID to set
    const browser = this.ctx.appData.browsers?.find((b) => this.idForBrowser(b) === browserId)

    if (!browser) {
      return
    }

    this.setActiveBrowser(browser)
  }
````

Since I don't see any use case where we want to set the active browser to be undefined.

File: packages/graphql/src/schemaTypes/objectTypes/gql-Mutation.ts
Comment (lmiller1990) [reply packages/graphql/src/schemaTypes/objectTypes/gql-Mutation.ts:235]: I think we probably just want to `return null` here - @tgriesser mentioned the convention when a mutation "fails" was just to `return null` if I recall corrrectly.

File: packages/data-context/src/actions/AppActions.ts
Comment (chrisbreiding) [reply packages/data-context/src/actions/AppActions.ts:15]: I guess I was opting to make it DRYer, since that check will have to go in both functions that call this.

File: packages/data-context/src/actions/AppActions.ts
Comment (chrisbreiding) [reply packages/data-context/src/actions/AppActions.ts:9]: It needs to be Bluebird since certain calling code uses Bluebird-specific APIs, like [here](https://github.com/cypress-io/cypress/blob/eeac2a28b572b5fbd99b5e295f341e04b73571b1/packages/server/lib/modes/run.js#L1557).

File: packages/graphql/src/schemaTypes/objectTypes/gql-Mutation.ts
Comment (chrisbreiding) [reply packages/graphql/src/schemaTypes/objectTypes/gql-Mutation.ts:235]: Done

File: packages/data-context/src/DataContext.ts
Comment (chrisbreiding) [reply packages/data-context/src/DataContext.ts:71]: Maybe in another PR.

Comment (tgriesser) [reply packages/data-context/src/DataContext.ts:71]: Oops, thought I had commented on this yesterday - yeah we can fix in another PR, some of the work I'm doing to refactor to add the CLI arguments in open mode tests will indirectly help clean this up, and some of the other changes to restructure some of the GraphQL types & internal state in #18709 will also help clean this yp.

File: packages/graphql/src/schemaTypes/objectTypes/gql-Mutation.ts
Comment (tgriesser) [reply packages/graphql/src/schemaTypes/objectTypes/gql-Mutation.ts:235]: For Query fields "failing", you'd want to return null, I think for mutations that are called in an invalid way it probably makes sense to throw b/c you have better control of the order of execution of those.

Comment (chrisbreiding) [reply packages/graphql/src/schemaTypes/objectTypes/gql-Mutation.ts:235]: So in this particular case, it should throw, right?

File: cli/__snapshots__/cli_spec.js
Comment (emilyrohrbough) [inline cli/__snapshots__/cli_spec.js:21]: by will attempt does that mean it'll fail if it's not found or will fallback to a different browser?

Comment (chrisbreiding) [reply cli/__snapshots__/cli_spec.js:21]: For `cypress run`, it will fail. For `cypress open`, it will show a message in the GUI and allow you to select a different browser.

Comment (emilyrohrbough) [reply cli/__snapshots__/cli_spec.js:21]: Sounds good. Might be something we can add/expand on in cypress-docs, but that'd be my first question as a user on how this would behave.

**Review by emilyrohrbough — APPROVED** [review]

**Comment by chrisbreiding** [comment]
Got approval from Ryan on the warning design.

File: packages/data-context/src/actions/AppActions.ts
Comment (ryanjwilke) [inline packages/data-context/src/actions/AppActions.ts:37]: My only concern with this is that "Choose a browser below" feels pretty repetitive given the title of the page is essentially this as well.

**Review by ryanjwilke — APPROVED** [review]
I left one comment that would be nice to address before releasing this, but if others disagree, then there's no need to block this PR over that from my perspective.

**Comment by tgriesser** [comment]
Working on fixing the issue that's causing the failure here

**Comment by tgriesser** [comment]
> I left one comment that would be nice to address before releasing this, but if others disagree, then there's no need to block this PR over that from my perspective.

@ryanjwilke Merging this PR to unblock separate work that's dependent on this landing to cleanup our testing

Probably worth making a note of your feedback and updating the wording here during our phase of testing/refining

**Pull request merged**
